test/unit/gateways/versa_pay_test.rb

```diff
@@ -0,0 +1,812 @@
+
+  def test_supported_card_types
+    assert_equal VersaPayGateway.supported_cardtypes, %i[visa master american_express discover]
+  end
+
+  def test_supported_countries
+    assert_equal VersaPayGateway.supported_countries, ['US']
+  end
+
+  def test_request_headers_building
+    gateway = VersaPayGateway.new(api_token: 'abc123', api_key: 'def456')
+    headers = gateway.send :request_headers
+
+    assert_equal 'application/json', headers['Content-Type']
+    assert_equal 'Basic YWJjMTIzOmRlZjQ1Ng==', headers['Authorization']
+  end
+
+  def test_build_order_request_url
+    action = :auth
+    assert_equal @gateway.send(:url, action), "#{@gateway.test_url}/api/gateway/v1/orders/auth"
+  end
+
+  def test_error_code_from_errors
+    # a HTTP 412 response structure
+    error = @gateway.send(:error_code_from, { 'success' => false, 'errors' => ['fund_address_unspecified'], 'response_code' => 999 })
+    assert_equal error, 'response_code: 999'
+  end
+
+  def test_error_code_from_gateway_error_code
+    error = @gateway.send(:error_code_from, declined_errors)
+    assert_equal error, 'gateway_error_code: 567.005 | response_code: 999'
+  end
+
+  def test_message_from_successful_purchase
+    message = @gateway.send(:message_from, @gateway.send(:parse, successful_purchase_response))
+    assert_equal message, 'Succeeded'
+  end
+
+  def test_message_from_failed_transaction_response
+    message = @gateway.send(:message_from, declined_errors)
+    assert_equal message, 'gateway_error_message: DECLINED | gateway_response_errors: [gateway - DECLINED]'
+  end
+
+  def test_message_from_failed_transaction_response_412
+    message = @gateway.send(:message_from, { 'success' => false, 'errors' => ['fund_address_unspecified'], 'response_code' => 999 })
+    assert_equal message, 'errors: fund_address_unspecified'
+  end
+
+  def test_successful_authorize
+    billing_address = @options[:billing_address]
+    @gateway.expects(:ssl_request).with(
+      :post,
+      'https://uat.versapay.com/api/gateway/v1/orders/auth',
+      all_of(
+        includes(
+          "amount_cents\":\"#{@amount}",
+          "email\":\"#{@options[:email]}\"",
+          'currency":"USD"',
+          "billing_name\":\"#{billing_address[:company]}\"",
+          "billing_address\":\"#{billing_address[:address1]}\"",
+          "billing_city\":\"#{billing_address[:city]}\"",
+          'billing_country":"CAN"',
+          "billing_email\":\"#{@options[:email]}\"",
+          "billing_telephone\":\"#{billing_address[:phone]}\"",
+          "name\":\"#{@credit_card.name}\"",
+          "expiry_month\":\"0#{@credit_card.month}\"",
+          "expiry_year\":#{@credit_card.year}",
+          "card_number\":\"#{@credit_card.number}\"",
+          "cvv\":\"#{@credit_card.verification_value}\""
+        )
+      ),
+      anything
+    ).
+      returns(successful_authorize_response)
+    @gateway.authorize(@amount, @credit_card, @options)
+  end
+
+  def test_successful_purchase
```

Comment:
💬 Comment

Given that this test is the same as the authorize one, maybe checking the request body is redundant.

Reply:
Right, I'll check the endpoint.